proposals/yield-in-try-catch.md

```diff
@@ -0,0 +1,388 @@
+
+Detailed notes:
+
+- The `yield` statement will be allowed in a `try` block.
+- The `yield` statement will be allowed in a `catch` block provided that:
+  - The `try` block does not contain a `finally` block.
+  - The `try` block does not have a nested `finally` block.
+
+### Dispose and finally
+
+A lesser known detail of iterators is that `finally` blocks can be executed as part of the `IDisposable.Dispose` implementation. The `Dispose` method has the same state machine implementation as the generated `MoveNext` except it only has the parts necessary for executing the `finally` blocks. That allows `Dispose` to _resume_ the method from the last suspend and execute the `finally` that were _active_ at the last suspend point.
+
+For example consider this code sample:
+
+```csharp
+var e = Iterator().GetEnumerator();
+e.MoveNext();
+e.Dispose();
+
+static T M<T>(T t) => t;
+
+static IEnumerable<int> Iterator()
+{
+    try
+    {
+        try
+        {
+            yield return M(1);
+            Console.WriteLine("After yield");
+        }
+        finally
+        {
+            Console.WriteLine("Finally Inner");
+        }
+    }
+    finally
+    {
+        Console.WriteLine("Finally Outer");
+    }
+}
+```
+
+This program will output:
+
+```cmd
+Finally Inner
+Finally Outer
+```
+
+The code generation for the `Dispose` method is meant to mirror the original `finally` structure as closely as possible. This includes execution of the code in the face of an exception during `Dispose`. This is achieved by refactoring the contents of the `finally` block into a method on the iterator and then having both `MoveNext` and `Dispose` generate the same `try / finally` structure and call into the methods.
+
+For example this is the `Dispose` method for the above iterator:
+
+```csharp
+[DebuggerHidden]
+void IDisposable.Dispose()
+{
+    int num = <>1__state;
+    if ((uint)(num - -4) > 1u && num != 1)
+    {
+        return;
+    }
+    try
+    {
+        if (num != -4 && num != 1)
+        {
+            return;
+        }
+        try
+        {
+        }
+        finally
+        {
+            <>m__Finally2();
+        }
+    }
+    finally
+    {
+        <>m__Finally1();
+    }
+}
+```
+
+This behavior is important to understand when considering the code generation for `try / catch` blocks.
+
+### Code generation yield inside try with catch
+
+The code generation for iterators that have `yield` inside `try` blocks must preserve the same exception semantics for the `catch` in both `MoveNext` and `Dispose`. To achieve this the compiler will take a similar approach to what it does for `finally` blocks.
+
+For every `catch` block where the `try` has a nested `try / finally` with `yield`:
+
+1. The contents of the `catch` block will be generated into a parameterless method on the iterator with a `void` return.
+2. The contents of the `when` clause will be generated into a parameterless method on the iterator with a `bool` return.
```

Comment:
Ah, I see that we don't need parameters for this extracted method after all. Makes sense :-)

Reply:
Yeah I realized it too once I started digging through a bunch of our generated code.